File: Tests/Patterns/LiteralTests.fs
namespace Tests.Patterns

open System
open Stringier.Patterns
open Microsoft.VisualStudio.TestTools.UnitTesting

[<TestClass>]
type LiteralTests() =

    [<TestMethod>]
    member _.``constructor`` () =
        let _charLiteral = p 'h'
        let _stringLiteral = p "hello"
        ()

    [<TestMethod>]
    member _.``consume string`` () =
        let hello = p "Hello"

        let mutable result = Result()
        
        result <- hello.Consume("Hello World!")
        ResultAssert.Captures("Hello", result)

        result <- hello.Consume("Hell")
        ResultAssert.Fails(result)

        result <- hello.Consume("Bacon")
        ResultAssert.Fails(result)

    [<TestMethod>]
    member _.``consume source`` () =
        let hello = p "Hello"
        let space = p ' '
        let world = p "World"

        let mutable result = Result()

        let mutable source = Source("Hello World")

        result <- hello.Consume(&source)
        ResultAssert.Captures("Hello", result)

        result <- space.Consume(&source)
        ResultAssert.Captures(" ", result)

        result <- world.Consume(&source)
        ResultAssert.Captures("World", result)

        source <- Source("Hello World")
        let helloWorld = hello >> space >> world

        result <- helloWorld.Consume(&source)
        ResultAssert.Captures("Hello World", result)

    [<TestMethod>]
    member _.``consume case-insensitive`` () =
        let pattern = "Hello"/=Compare.CaseInsensitive >> ' ' >> "World"/=Compare.CaseInsensitive
        
        ResultAssert.Captures("HELLO WORLD", pattern.Consume("HELLO WORLD"))
        ResultAssert.Captures("hello world", pattern.Consume("hello world"))

    [<TestMethod>]
    member _.``neglect string`` () =
        let pattern = negate "Hello"
        ResultAssert.Fails(pattern.Consume("Hello"))
        ResultAssert.Fails(pattern.Consume("Hello!"))
        ResultAssert.Fails(pattern.Consume("Hello."))
        ResultAssert.Captures("Oh no", pattern.Consume("Oh no!"))
        ResultAssert.Captures("Oh no", pattern.Consume("Oh no?"))

File: Benchmarks.Tests/LiteralTests.cs
using System;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Pidgin;
using Sprache;
using static Benchmarks.Patterns.LiteralBenchmarks;

namespace Benchmarks {
	[TestClass]
	public class LiteralTests {
		[TestMethod]
		public void Pidgin() {
			Result<Char, String> result;

			result = pidgin.Parse("Hello");
			Assert.IsTrue(result.Success);
			Assert.AreEqual("Hello", result.Value);

			result = pidgin.Parse("World");
			Assert.IsFalse(result.Success);

			result = pidgin.Parse("H");
			Assert.IsFalse(result.Success);

			result = pidgin.Parse("Failure");
			Assert.IsFalse(result.Success);
		}

		[TestMethod]
		public void Sprache() {
			IResult<String> result;

			result = sprache.TryParse("Hello");
			Assert.IsTrue(result.WasSuccessful);
			Assert.AreEqual("Hello", result.Value);

			result = sprache.TryParse("World");
			Assert.IsFalse(result.WasSuccessful);

			result = sprache.TryParse("H");
			Assert.IsFalse(result.WasSuccessful);

			result = sprache.TryParse("Failure");
			Assert.IsFalse(result.WasSuccessful);
		}
	}
}
